[quant][graphmode][fx] Separate handling Copy operator to a helper function

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54928 [quant][graphmode][fx][refactor] Quantize by Use of a Tensor instead of Tensor
* #54924 [quant][graphmode][fx] Produce torch.cat instead of torch.ops.quantized.cat
* #54813 [quant][graphmode][fx] Optimize cat
* #54812 [nn] Add allow_duplicate option for named_modules
* #54733 [quant][graphmode][fx][refactor] Factor out insert_observers_for_model to a separate function
* **#54644 [quant][graphmode][fx] Separate handling Copy operator to a helper function**
* #54643 [quant][fx][graphmode][refactor] Change activation_post_process_map to track the observer name instead
* #54073 [quant][graphmode][refactor] Remove reduandent code
* #53586 [quant][graphmode][fx] Add support for one value being quantized with different qconfigs

Summary:
Previously we special case copy operator in normal insert observer code, this PR tries to split the
special case logic to a separate function and keep the rest of the code clean.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27314678](https://our.internmc.facebook.com/intern/diff/D27314678)

jerryzh168 committed:
[quant][graphmode][fx] Separate handling Copy operator to a helper function

Summary:
Previously we special case copy operator in normal insert observer code, this PR tries to split the
special case logic to a separate function and keep the rest of the code clean.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Separate handling Copy operator to a helper function"


Summary:
Previously we special case copy operator in normal insert observer code, this PR tries to split the
special case logic to a separate function and keep the rest of the code clean.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27314678](https://our.internmc.facebook.com/intern/diff/D27314678)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Separate handling Copy operator to a helper function"


Summary:
Previously we special case copy operator in normal insert observer code, this PR tries to split the
special case logic to a separate function and keep the rest of the code clean.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27314678](https://our.internmc.facebook.com/intern/diff/D27314678)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Separate handling Copy operator to a helper function"


Summary:
Previously we special case copy operator in normal insert observer code, this PR tries to split the
special case logic to a separate function and keep the rest of the code clean.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27314678](https://our.internmc.facebook.com/intern/diff/D27314678)

[ghstack-poisoned]

vkuzo commented on torch/quantization/fx/quantize.py:
```diff
@@ -620,6 +685,16 @@ def load_arg(a):
                 observed_graph, load_arg)
 
 
+        self.modules = dict(model.named_modules())
```
Would it make more sense to pass the top level model around and look up the right module every time?  That way there would be no additional state variables created, so it might be easier to think about the code.  Note: I didn't think through this deeply, more of a question.

jerryzh168 replied:
makes sense, I think we should be able to remove self.modules and replace the access
```
self.modules[node.target]
```
with
```
getattr(model, node.target)
```
will do this in a separate refactor PR

jerryzh168 replied:
just tried this, I think we cannot do it, getattr(model, '') or getattr(model, 'conv.0') is not going to work

vkuzo commented on torch/quantization/fx/utils.py:
```diff
@@ -339,7 +340,7 @@ def create_qparam_nodes(quantizer: QuantizerCls, node_name: str, scale: Any, zer
 
 
-def all_node_args_have_no_tensors(node: Node) -> bool:
+def all_node_args_have_no_tensors(node: Node, modules: Dict[str, torch.nn.Module]) -> bool:
```
would it be possible to pass around the parent module instead, and look up the module dynamically every time?  This avoids the need for one more custom stateful container.  I think the performance concern is ok (this shouldn't be the bottleneck) if we get lower code complexity in return.

jerryzh168 replied:
not sure I understand, can you clarify with code?

vkuzo replied:
oh, it's the same thing from the above comment, your suggestion of doing `getattr(model, node.target)` sounds great, separate PR sounds great

vkuzo commented on torch/quantization/fx/quantize.py:
```diff
@@ -620,6 +685,16 @@ def load_arg(a):
+            standalone_module_classes, custom_module_classes)
+        quants = self._find_quants(observed_graph, self.modules, matches)
+
+        observed_graph = handle_copy_nodes(observed_graph, matches, quants, self.activation_post_process_map, self.modules)
```
one source of confusion is that L620:L685 are inlined, but this line is in a separate function.  Thoughts about either inlining both or having both in separate functions, just so it's consistent on where this style of "go through all the nodes and create a new graph" code is happening?

jerryzh168 replied:
sounds good

jerryzh168 replied:
will do this in a follow up PR

jerryzh168 committed:
Update on "[quant][graphmode][fx] Separate handling Copy operator to a helper function"


Summary:
Previously we special case copy operator in normal insert observer code, this PR tries to split the
special case logic to a separate function and keep the rest of the code clean.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27314678](https://our.internmc.facebook.com/intern/diff/D27314678)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Separate handling Copy operator to a helper function"


Summary:
Previously we special case copy operator in normal insert observer code, this PR tries to split the
special case logic to a separate function and keep the rest of the code clean.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27314678](https://our.internmc.facebook.com/intern/diff/D27314678)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Separate handling Copy operator to a helper function"


Summary:
Previously we special case copy operator in normal insert observer code, this PR tries to split the
special case logic to a separate function and keep the rest of the code clean.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27314678](https://our.internmc.facebook.com/intern/diff/D27314678)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Separate handling Copy operator to a helper function"


Summary:
Previously we special case copy operator in normal insert observer code, this PR tries to split the
special case logic to a separate function and keep the rest of the code clean.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27314678](https://our.internmc.facebook.com/intern/diff/D27314678)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Separate handling Copy operator to a helper function"


Summary:
Previously we special case copy operator in normal insert observer code, this PR tries to split the
special case logic to a separate function and keep the rest of the code clean.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27314678](https://our.internmc.facebook.com/intern/diff/D27314678)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Separate handling Copy operator to a helper function"


Summary:
Previously we special case copy operator in normal insert observer code, this PR tries to split the
special case logic to a separate function and keep the rest of the code clean.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27314678](https://our.internmc.facebook.com/intern/diff/D27314678)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Separate handling Copy operator to a helper function"


Summary:
Previously we special case copy operator in normal insert observer code, this PR tries to split the
special case logic to a separate function and keep the rest of the code clean.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27314678](https://our.internmc.facebook.com/intern/diff/D27314678)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Separate handling Copy operator to a helper function"


Summary:
Previously we special case copy operator in normal insert observer code, this PR tries to split the
special case logic to a separate function and keep the rest of the code clean.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27314678](https://our.internmc.facebook.com/intern/diff/D27314678)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Separate handling Copy operator to a helper function"


Summary:
Previously we special case copy operator in normal insert observer code, this PR tries to split the
special case logic to a separate function and keep the rest of the code clean.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27314678](https://our.internmc.facebook.com/intern/diff/D27314678)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Separate handling Copy operator to a helper function"


Summary:
Previously we special case copy operator in normal insert observer code, this PR tries to split the
special case logic to a separate function and keep the rest of the code clean.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27314678](https://our.internmc.facebook.com/intern/diff/D27314678)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Separate handling Copy operator to a helper function"


Summary:
Previously we special case copy operator in normal insert observer code, this PR tries to split the
special case logic to a separate function and keep the rest of the code clean.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27314678](https://our.internmc.facebook.com/intern/diff/D27314678)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Separate handling Copy operator to a helper function"


Summary:
Previously we special case copy operator in normal insert observer code, this PR tries to split the
special case logic to a separate function and keep the rest of the code clean.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27314678](https://our.internmc.facebook.com/intern/diff/D27314678)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Separate handling Copy operator to a helper function"


Summary:
Previously we special case copy operator in normal insert observer code, this PR tries to split the
special case logic to a separate function and keep the rest of the code clean.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27314678](https://our.internmc.facebook.com/intern/diff/D27314678)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Separate handling Copy operator to a helper function"


Summary:
Previously we special case copy operator in normal insert observer code, this PR tries to split the
special case logic to a separate function and keep the rest of the code clean.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27314678](https://our.internmc.facebook.com/intern/diff/D27314678)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Separate handling Copy operator to a helper function"


Summary:
Previously we special case copy operator in normal insert observer code, this PR tries to split the
special case logic to a separate function and keep the rest of the code clean.

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27314678](https://our.internmc.facebook.com/intern/diff/D27314678)

[ghstack-poisoned]
